packages/lwc-language-server/src/component-indexer.ts

```diff
@@ -0,0 +1,135 @@
+import Tag from './tag';
+import * as path from 'path';
+import { shared } from '@salesforce/lightning-lsp-common';
+import { Entry, sync } from 'fast-glob';
+import * as fsExtra from 'fs-extra';
+import { join } from 'path';
+import { snakeCase } from 'change-case';
+import BaseIndexer from './base-indexer';
+
+const { detectWorkspaceHelper, WorkspaceType } = shared;
+const CUSTOM_COMPONENT_INDEX_FILE = '.sfdx/indexes/lwc/custom-components.json';
+const componentPrefixRegex = new RegExp(/^(?<type>c|lightning|interop){0,1}(?<delimiter>:|-{0,1})(?<name>[\w\-]+)$/);
+
+type ComponentIndexerAttributes = {
+    workspaceRoot: string;
+};
+
+export enum DelimiterType {
+    Aura = ':',
+    LWC = '-',
+}
+
+export function tagEqualsFile(tag: Tag, entry: Entry): boolean {
+    return tag.file === entry.path && tag.updatedAt?.getTime() === entry.stats?.mtime.getTime();
+}
+
+export function unIndexedFiles(entries: Entry[], tags: Tag[]) {
+    return entries.filter(entry => !tags.some(tag => tagEqualsFile(tag, entry)));
+}
+
+export default class ComponentIndexer extends BaseIndexer {
+    readonly workspaceType: number;
+    readonly tags: Map<string, Tag> = new Map();
+
+    constructor(attributes: ComponentIndexerAttributes) {
+        super(attributes);
+        this.workspaceType = detectWorkspaceHelper(attributes.workspaceRoot);
+    }
+
+    get componentEntries(): Entry[] {
+        let files: Entry[] = [];
+        switch (this.workspaceType) {
+            case WorkspaceType.SFDX:
+                files = sync(path.join(this.workspaceRoot, this.sfdxPackageDirsPattern, '**/*/lwc/**/*.js'), {
+                    stats: true,
+                });
+                return files.filter((item: Entry): boolean => {
+                    const data = path.parse(item.path);
+                    return data.dir.endsWith(data.name);
+                });
+            default:
+                // For CORE_ALL and CORE_PARTIAL
+                files = sync(path.join(this.workspaceRoot, '**/*/modules/**/*.js'), {
+                    stats: true,
+                });
+                return files.filter((item: Entry): boolean => {
+                    const data = path.parse(item.path);
+                    return data.dir.endsWith(data.name);
+                });
+        }
+    }
+
+    get customData(): Tag[] {
+        return Array.from(this.tags.values());
+    }
+
+    findTagByName(query: string): Tag | null {
+        const matches = componentPrefixRegex.exec(query);
+        const { type, name } = matches.groups;
```

Comment:
This should be guarded in case of null matches. It produces an error for certain onDefinition() calls.